Don't enable RUST_BOOTSTRAP in build.rs

The crate is currently "usable" on stable rust, despite using nightly-only feature, by enabling RUST_BOOTSTRAP=1 in build.rs. This is a **very** bad idea. This flag is only meant to bootstrap the compiler itself, and any other use of it is unsupported and might lead to the code breaking if/when the API changes. This is not unheard of: Nightly features aren't finalized and often changes shape prior to stabilization.

The currently used features are also not very stable, const_generics and const_fn can easily cause ICEs in the compiler or soundness issues in the final binary, due to implementation bugs.

IMO, this crate shouldn't abuse RUST_BOOTSTRAP like this. The correct way is to tell the users to use a nightly compiler. If that's not possible, the use of RUST_BOOTSTRAP should at least be called out in the README so users can make an informed decision about using this crate.

So, this answer was a bit disappointing to hear, as it makes it pretty hard to rely on obfstr on production code. I have forked obfstr 1.1 and, with the use of `generic_array` managed to make something working reasonably well on stable without the need for any nightly feature. See https://github.com/roblabla/obfstr/commit/195db0c91af3f7fe93cbbb92126e52c61c4517ea

Only real downside is that it makes `obfconst` slightly less ergonomic to use:

```rust
static GSTR: obfstr::ObfString<[u8; 10]> = obfstr::obfconst!("Hello 🌍");
use generic_array::{GenericArray, typenum::consts::*};
static GSTR: obfstr::ObfString<GenericArray<u8, U10>> = obfstr::obfconst!("Hello 🌍");
assert_eq!(GSTR.decrypt(0).as_str(), "Hello 🌍");
```

And, as with anything relying on generic-array, the errors that occur when giving the wrong output size are pretty bad if you don't know how to read them:

```

---- src/lib.rs - obfconst (line 72) stdout ----
error[E0308]: mismatched types
 --> src/lib.rs:74:56
  |
5 | static GSTR: obfstr::ObfString<GenericArray<u8, U9>> = obfstr::obfconst!("Hello 🌍");
  |                                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `obfstr::typenum::B0`, found struct `obfstr::typenum::B1`
  |
  = note: expected struct `obfstr::ObfString<obfstr::GenericArray<_, obfstr::typenum::UInt<obfstr::typenum::UInt<_, obfstr::typenum::B0>, obfstr::typenum::B1>>>`
             found struct `obfstr::ObfString<obfstr::GenericArray<_, obfstr::typenum::UInt<obfstr::typenum::UInt<_, obfstr::typenum::B1>, obfstr::typenum::B0>>>`
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error
```

So my question is this: are you open to moving to such a solution? If not, I guess I'll publish my fork on crates.io with another name (like `obfuscate` or something).

--- a/build.rs
+++ b/build.rs
@@ -1,4 +1,0 @@
-fn main() {
-	// Please I need to be generic over _all_ arrays._.;
-	println!("cargo:rustc-env=RUSTC_BOOTSTRAP=1");
-}

--- a/impl/src/lib.rs
+++ b/impl/src/lib.rs
@@ -12,20 +12,20 @@ use proc_macro::*;
 /// by a `macro_rules!` expansion.
 /// See https://github.com/rust-lang/rust/issues/72545 for background
 fn ignore_groups(mut input: TokenStream) -> TokenStream {
-    loop {
-        let mut tokens = input.clone().into_iter();
-        if let Some(TokenTree::Group(group)) = tokens.next() {
-            if group.delimiter() == Delimiter::None {
-                input = group.stream();
-                continue;
-            }
-        }
-        return input;
-    }
+	loop {
+		let mut tokens = input.clone().into_iter();
+		if let Some(TokenTree::Group(group)) = tokens.next() {
+			if group.delimiter() == Delimiter::None {
+				input = group.stream();
+				continue;
+			}
+		}
+		return input;
+	}
 }
 
 fn strip_groups(token: TokenTree) -> TokenTree {
-    ignore_groups(vec![token].into_iter().collect()).into_iter().next().unwrap()
+	ignore_groups(vec![token].into_iter().collect()).into_iter().next().unwrap()
 }
 
 #[cfg(feature = "rand")]
@@ -37,7 +37,7 @@ pub fn obfstr_attribute(args: TokenStream, input: TokenStream) -> TokenStream {
 
 #[cfg(feature = "rand")]
 fn strlen_impl(mut input: TokenStream) -> TokenStream {
-    input = ignore_groups(input);
+	input = ignore_groups(input);
 	if let Some(TokenTree::Literal(literal)) = input.into_iter().next() {
 		let s = string_parse(literal);
 		TokenStream::from(TokenTree::Literal(Literal::usize_suffixed(s.len())))
@@ -48,7 +48,7 @@ fn strlen_impl(mut input: TokenStream) -> TokenStream {
 }
 #[cfg(feature = "rand")]
 fn obfstr_impl(mut input: TokenStream) -> TokenStream {
-    input = ignore_groups(input);
+	input = ignore_groups(input);
 	let mut tt = input.into_iter();
 	let mut token = tt.next();
 
@@ -195,14 +195,18 @@ fn string_parse(input: Literal) -> String {
 
 //----------------------------------------------------------------
 
-#[proc_macro_attribute]
-pub fn wide_attribute(args: TokenStream, input: TokenStream) -> TokenStream {
-	drop(args);
-	replace_macro(input, "_wide_", wide_impl)
-}
-
-fn wide_impl(mut input: TokenStream) -> TokenStream {
-    input = ignore_groups(input);
+/// Wide string literal, returns an array of words.
+///
+/// The type of the returned literal is `&'static [u16; LEN]`.
+///
+/// ```
+/// # use obfstr_impl as obfstr;
+/// let expected = &['W' as u16, 'i' as u16, 'd' as u16, 'e' as u16, 0];
+/// assert_eq!(obfstr::wide!("Wide\0"), expected);
+/// ```
+#[proc_macro]
+pub fn wide(input: TokenStream) -> TokenStream {
+	let input = ignore_groups(input);
 	// Parse the input as a single string literal
 	let mut iter = input.into_iter();
 	let string = match iter.next() {
@@ -229,46 +233,47 @@ fn wide_impl(mut input: TokenStream) -> TokenStream {
 
 //----------------------------------------------------------------
 
+/// Compiletime random number generator.
+///
+/// Every time the code is compiled, a new random number literal is generated.
+/// Recompilation (and thus regeneration of the number) is not triggered automatically.
+///
+/// Supported types are `u8`, `u16`, `u32`, `u64`, `usize`, `i8`, `i16`, `i32`, `i64`, `isize`, `bool`, `f32` and `f64`.
+///
+/// ```
+/// # use obfstr_impl as obfstr;
+/// const RND: i32 = obfstr::random!(u8) as i32;
+/// assert!(RND >= 0 && RND <= 255);
+/// ```
 #[cfg(feature = "rand")]
-#[proc_macro_attribute]
-pub fn random_attribute(args: TokenStream, input: TokenStream) -> TokenStream {
-	drop(args);
-	replace_macro(input, "_random_", random_impl)
-}
-
-#[cfg(feature = "rand")]
-fn random_impl(mut input: TokenStream) -> TokenStream {
-    input = ignore_groups(input);
+#[proc_macro]
+pub fn random(input: TokenStream) -> TokenStream {
+	let input = ignore_groups(input);
 	let mut tt = input.into_iter();
 	match tt.next() {
 		Some(TokenTree::Ident(ident)) => {
 			if let Some(tt) = tt.next() {
 				panic!("unexpected token: `{}`", tt);
 			}
-			random_parse(ident).into()
+			TokenTree::Literal(match &*ident.to_string() {
+				"u8" => Literal::u8_suffixed(rand::random::<u8>()),
+				"u16" => Literal::u16_suffixed(rand::random::<u16>()),
+				"u32" => Literal::u32_suffixed(rand::random::<u32>()),
+				"u64" => Literal::u64_suffixed(rand::random::<u64>()),
+				"usize" => Literal::usize_suffixed(rand::random::<usize>()),
+				"i8" => Literal::i8_suffixed(rand::random::<i8>()),
+				"i16" => Literal::i16_suffixed(rand::random::<i16>()),
+				"i32" => Literal::i32_suffixed(rand::random::<i32>()),
+				"i64" => Literal::i64_suffixed(rand::random::<i64>()),
+				"isize" => Literal::isize_suffixed(rand::random::<isize>()),
+				"f32" => Literal::f32_suffixed(rand::random::<f32>()),
+				"f64" => Literal::f64_suffixed(rand::random::<f64>()),
+				s => panic!("unsupported type: `{}`", s),
+			}).into()
 		},
 		Some(tt) => panic!("expected a primitive name: `{}`", tt),
 		None => panic!("expected a primitive name"),
 	}
-}
-
-#[cfg(feature = "rand")]
-fn random_parse(input: Ident) -> TokenTree {
-	match &*input.to_string() {
-		"u8" => Literal::u8_suffixed(rand::random::<u8>()),
-		"u16" => Literal::u16_suffixed(rand::random::<u16>()),
-		"u32" => Literal::u32_suffixed(rand::random::<u32>()),
-		"u64" => Literal::u64_suffixed(rand::random::<u64>()),
-		"usize" => Literal::usize_suffixed(rand::random::<usize>()),
-		"i8" => Literal::i8_suffixed(rand::random::<i8>()),
-		"i16" => Literal::i16_suffixed(rand::random::<i16>()),
-		"i32" => Literal::i32_suffixed(rand::random::<i32>()),
-		"i64" => Literal::i64_suffixed(rand::random::<i64>()),
-		"isize" => Literal::isize_suffixed(rand::random::<isize>()),
-		"f32" => Literal::f32_suffixed(rand::random::<f32>()),
-		"f64" => Literal::f64_suffixed(rand::random::<f64>()),
-		s => panic!("unsupported type: `{}`", s),
-	}.into()
 }
 
 //----------------------------------------------------------------

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -7,6 +7,7 @@ Compiletime string literal obfuscation.
 // Reexport these because reasons...
 #[doc(hidden)]
 pub use obfstr_impl::*;
+pub use obfstr_impl::wide;
 
 #[cfg(feature = "rand")]
 pub use cfgd::*;
@@ -14,6 +15,7 @@ pub use cfgd::*;
 mod cfgd {
 
 use core::{char, fmt, mem, ops, ptr, slice, str};
+pub use obfstr_impl::random;
 
 /// Compiletime string literal obfuscation, returns a borrowed temporary and may not escape the statement it was used in.
 ///
@@ -90,25 +92,6 @@ macro_rules! unsafe_obfstr {
 		#[$crate::obfstr_attribute]
 		const S: $crate::ObfString<[u8; _strlen_!($string)]> = $crate::ObfString::new(_obfstr_!($string));
 		S.decrypt($crate::random!(usize) & 0xffff).unsafe_as_static_str()
-	}};
-}
-
-/// Compiletime random number generator.
-///
-/// Every time the code is compiled, a new random number literal is generated.
-/// Recompilation (and thus regeneration of the number) is not triggered automatically.
-///
-/// Supported types are `u8`, `u16`, `u32`, `u64`, `usize`, `i8`, `i16`, `i32`, `i64`, `isize`, `bool`, `f32` and `f64`.
-///
-/// ```
-/// const RND: i32 = obfstr::random!(u8) as i32;
-/// assert!(RND >= 0 && RND <= 255);
-/// ```
-#[macro_export]
-macro_rules! random {
-	($ty:ident) => {{
-		#[$crate::random_attribute]
-		const N: $ty = _random_!($ty); N
 	}};
 }
 
@@ -326,20 +309,4 @@ impl<A: AsRef<[u16]>> fmt::Display for WObfBuffer<A> {
 		Ok(())
 	}
 }
-}
-
-/// Wide string literal, returns an array of words.
-///
-/// The type of the returned literal is `&'static [u16; LEN]`.
-///
-/// ```
-/// let expected = &['W' as u16, 'i' as u16, 'd' as u16, 'e' as u16];
-/// assert_eq!(obfstr::wide!("Wide"), expected);
-/// ```
-#[macro_export]
-macro_rules! wide {
-	($s:literal) => {{
-		#[$crate::wide_attribute]
-		const W: &[u16] = _wide_!($s); W
-	}};
 }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -15,6 +15,7 @@ pub use cfgd::*;
 mod cfgd {
 
 use core::{char, fmt, mem, ops, ptr, slice, str};
+
 pub use obfstr_impl::random;
 
 /// Compiletime string literal obfuscation, returns a borrowed temporary and may not escape the statement it was used in.
@@ -124,7 +125,7 @@ impl<A> ObfString<A> {
 		ObfString { key, data }
 	}
 }
-impl<A: AsRef<[u8]> + AsMut<[u8]>> ObfString<A> {
+impl<A: AsRef<[u8]> + AsMut<[u8]> + Clone> ObfString<A> {
 	/// Decrypts the obfuscated string and returns the buffer.
 	///
 	/// The `x` argument should be a compiletime random 16-bit value.
@@ -132,21 +133,21 @@ impl<A: AsRef<[u8]> + AsMut<[u8]>> ObfString<A> {
 	#[inline(always)]
 	pub fn decrypt(&self, x: usize) -> ObfBuffer<A> {
 		unsafe {
-			let mut buffer = ObfBuffer::<A>::uninit();
+			let mut buffer = self.data.clone();
 			let data = self.data.as_ref();
 			let src = data.as_ptr() as usize - data.len() * XREF_SHIFT;
 			let f: unsafe fn(&mut [u8], usize) = mem::transmute(ptr::read_volatile(&(decryptbuf as usize + x)) - x);
-			f(buffer.0.as_mut(), src);
-			buffer
+			f(buffer.as_mut(), src);
+			ObfBuffer(buffer)
 		}
 	}
 }
-impl<A: AsRef<[u8]> + AsMut<[u8]>> fmt::Debug for ObfString<A> {
+impl<A: AsRef<[u8]> + AsMut<[u8]> + Clone> fmt::Debug for ObfString<A> {
 	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
 		self.decrypt(random!(usize) & 0xffff).fmt(f)
 	}
 }
-impl<A: AsRef<[u8]> + AsMut<[u8]>> fmt::Display for ObfString<A> {
+impl<A: AsRef<[u8]> + AsMut<[u8]> + Clone> fmt::Display for ObfString<A> {
 	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
 		self.decrypt(random!(usize) & 0xffff).fmt(f)
 	}
@@ -167,10 +168,6 @@ unsafe fn decryptbuf(dest: &mut [u8], src: usize) {
 #[repr(transparent)]
 pub struct ObfBuffer<A>(A);
 impl<A: AsRef<[u8]>> ObfBuffer<A> {
-	#[allow(deprecated)]
-	unsafe fn uninit() -> Self {
-		mem::uninitialized()
-	}
 	#[inline]
 	pub fn as_str(&self) -> &str {
 		#[cfg(debug_assertions)]
@@ -226,7 +223,7 @@ impl<A> WObfString<A> {
 		WObfString { key, data }
 	}
 }
-impl<A: AsRef<[u16]> + AsMut<[u16]>> WObfString<A> {
+impl<A: AsRef<[u16]> + AsMut<[u16]> + Clone> WObfString<A> {
 	/// Decrypts the obfuscated wide string and returns the buffer.
 	///
 	/// The `x` argument should be a compiletime random 16-bit value.
@@ -234,21 +231,21 @@ impl<A: AsRef<[u16]> + AsMut<[u16]>> WObfString<A> {
 	#[inline(always)]
 	pub fn decrypt(&self, x: usize) -> WObfBuffer<A> {
 		unsafe {
-			let mut buffer = WObfBuffer::<A>::uninit();
+			let mut buffer = self.data.clone();
 			let data = self.data.as_ref();
 			let src = data.as_ptr() as usize - data.len() * XREF_SHIFT;
 			let f: unsafe fn(&mut [u16], usize) = mem::transmute(ptr::read_volatile(&(wdecryptbuf as usize + x)) - x);
-			f(buffer.0.as_mut(), src);
-			buffer
+			f(buffer.as_mut(), src);
+			WObfBuffer(buffer)
 		}
 	}
 }
-impl<A: AsRef<[u16]> + AsMut<[u16]>> fmt::Debug for WObfString<A> {
+impl<A: AsRef<[u16]> + AsMut<[u16]> + Clone> fmt::Debug for WObfString<A> {
 	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
 		self.decrypt(random!(usize) & 0xffff).fmt(f)
 	}
 }
-impl<A: AsRef<[u16]> + AsMut<[u16]>> fmt::Display for WObfString<A> {
+impl<A: AsRef<[u16]> + AsMut<[u16]> + Clone> fmt::Display for WObfString<A> {
 	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
 		self.decrypt(random!(usize) & 0xffff).fmt(f)
 	}
@@ -269,10 +266,6 @@ unsafe fn wdecryptbuf(dest: &mut [u16], src: usize) {
 #[repr(transparent)]
 pub struct WObfBuffer<A>(A);
 impl<A: AsRef<[u16]>> WObfBuffer<A> {
-	#[allow(deprecated)]
-	unsafe fn uninit() -> Self {
-		mem::uninitialized()
-	}
 	#[inline]
 	pub fn as_wide(&self) -> &[u16] {
 		self.0.as_ref()

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -132,14 +132,14 @@ impl<A: AsRef<[u8]> + AsMut<[u8]> + Clone> ObfString<A> {
 	/// It is used to obfuscate the underlying call to the decrypt routine.
 	#[inline(always)]
 	pub fn decrypt(&self, x: usize) -> ObfBuffer<A> {
+		let mut buffer = self.data.clone();
+		let data = self.data.as_ref();
+		let src = data.as_ptr() as usize - data.len() * XREF_SHIFT;
 		unsafe {
-			let mut buffer = self.data.clone();
-			let data = self.data.as_ref();
-			let src = data.as_ptr() as usize - data.len() * XREF_SHIFT;
 			let f: unsafe fn(&mut [u8], usize) = mem::transmute(ptr::read_volatile(&(decryptbuf as usize + x)) - x);
 			f(buffer.as_mut(), src);
-			ObfBuffer(buffer)
 		}
+		ObfBuffer(buffer)
 	}
 }
 impl<A: AsRef<[u8]> + AsMut<[u8]> + Clone> fmt::Debug for ObfString<A> {
@@ -230,14 +230,14 @@ impl<A: AsRef<[u16]> + AsMut<[u16]> + Clone> WObfString<A> {
 	/// It is used to obfuscate the underlying call to the decrypt routine.
 	#[inline(always)]
 	pub fn decrypt(&self, x: usize) -> WObfBuffer<A> {
+		let mut buffer = self.data.clone();
+		let data = self.data.as_ref();
+		let src = data.as_ptr() as usize - data.len() * XREF_SHIFT;
 		unsafe {
-			let mut buffer = self.data.clone();
-			let data = self.data.as_ref();
-			let src = data.as_ptr() as usize - data.len() * XREF_SHIFT;
 			let f: unsafe fn(&mut [u16], usize) = mem::transmute(ptr::read_volatile(&(wdecryptbuf as usize + x)) - x);
 			f(buffer.as_mut(), src);
-			WObfBuffer(buffer)
 		}
+		WObfBuffer(buffer)
 	}
 }
 impl<A: AsRef<[u16]> + AsMut<[u16]> + Clone> fmt::Debug for WObfString<A> {

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -3,7 +3,6 @@ Compiletime string literal obfuscation.
 !*/
 
 #![no_std]
-#![feature(fixed_size_array)]
 
 // Reexport these because reasons...
 #[doc(hidden)]
@@ -15,7 +14,6 @@ pub use cfgd::*;
 mod cfgd {
 
 use core::{char, fmt, mem, ops, ptr, slice, str};
-use core::array::FixedSizeArray;
 
 /// Compiletime string literal obfuscation, returns a borrowed temporary and may not escape the statement it was used in.
 ///
@@ -143,7 +141,7 @@ impl<A> ObfString<A> {
 		ObfString { key, data }
 	}
 }
-impl<A: FixedSizeArray<u8>> ObfString<A> {
+impl<A: AsRef<[u8]> + AsMut<[u8]>> ObfString<A> {
 	/// Decrypts the obfuscated string and returns the buffer.
 	///
 	/// The `x` argument should be a compiletime random 16-bit value.
@@ -152,20 +150,20 @@ impl<A: FixedSizeArray<u8>> ObfString<A> {
 	pub fn decrypt(&self, x: usize) -> ObfBuffer<A> {
 		unsafe {
 			let mut buffer = ObfBuffer::<A>::uninit();
-			let data = self.data.as_slice();
+			let data = self.data.as_ref();
 			let src = data.as_ptr() as usize - data.len() * XREF_SHIFT;
 			let f: unsafe fn(&mut [u8], usize) = mem::transmute(ptr::read_volatile(&(decryptbuf as usize + x)) - x);
-			f(buffer.0.as_mut_slice(), src);
+			f(buffer.0.as_mut(), src);
 			buffer
 		}
 	}
 }
-impl<A: FixedSizeArray<u8>> fmt::Debug for ObfString<A> {
+impl<A: AsRef<[u8]> + AsMut<[u8]>> fmt::Debug for ObfString<A> {
 	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
 		self.decrypt(random!(usize) & 0xffff).fmt(f)
 	}
 }
-impl<A: FixedSizeArray<u8>> fmt::Display for ObfString<A> {
+impl<A: AsRef<[u8]> + AsMut<[u8]>> fmt::Display for ObfString<A> {
 	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
 		self.decrypt(random!(usize) & 0xffff).fmt(f)
 	}
@@ -185,7 +183,7 @@ unsafe fn decryptbuf(dest: &mut [u8], src: usize) {
 #[derive(Copy, Clone)]
 #[repr(transparent)]
 pub struct ObfBuffer<A>(A);
-impl<A: FixedSizeArray<u8>> ObfBuffer<A> {
+impl<A: AsRef<[u8]>> ObfBuffer<A> {
 	#[allow(deprecated)]
 	unsafe fn uninit() -> Self {
 		mem::uninitialized()
@@ -193,7 +191,7 @@ impl<A: FixedSizeArray<u8>> ObfBuffer<A> {
 	#[inline]
 	pub fn as_str(&self) -> &str {
 		#[cfg(debug_assertions)]
-		return str::from_utf8(self.0.as_slice()).unwrap();
+		return str::from_utf8(self.0.as_ref()).unwrap();
 		#[cfg(not(debug_assertions))]
 		return unsafe { str::from_utf8_unchecked(self.0.as_slice()) };
 	}
@@ -204,24 +202,24 @@ impl<A: FixedSizeArray<u8>> ObfBuffer<A> {
 		unsafe { &*(self.as_str() as *const _) }
 	}
 }
-impl<A: FixedSizeArray<u8>> ops::Deref for ObfBuffer<A> {
+impl<A: AsRef<[u8]>> ops::Deref for ObfBuffer<A> {
 	type Target = str;
 	#[inline]
 	fn deref(&self) -> &str {
 		self.as_str()
 	}
 }
-impl<A: FixedSizeArray<u8>> AsRef<str> for ObfBuffer<A> {
+impl<A: AsRef<[u8]>> AsRef<str> for ObfBuffer<A> {
 	fn as_ref(&self) -> &str {
 		self.as_str()
 	}
 }
-impl<A: FixedSizeArray<u8>> fmt::Debug for ObfBuffer<A> {
+impl<A: AsRef<[u8]>> fmt::Debug for ObfBuffer<A> {
 	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
 		self.as_str().fmt(f)
 	}
 }
-impl<A: FixedSizeArray<u8>> fmt::Display for ObfBuffer<A> {
+impl<A: AsRef<[u8]>> fmt::Display for ObfBuffer<A> {
 	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
 		self.as_str().fmt(f)
 	}
@@ -245,7 +243,7 @@ impl<A> WObfString<A> {
 		WObfString { key, data }
 	}
 }
-impl<A: FixedSizeArray<u16>> WObfString<A> {
+impl<A: AsRef<[u16]> + AsMut<[u16]>> WObfString<A> {
 	/// Decrypts the obfuscated wide string and returns the buffer.
 	///
 	/// The `x` argument should be a compiletime random 16-bit value.
@@ -254,20 +252,20 @@ impl<A: FixedSizeArray<u16>> WObfString<A> {
 	pub fn decrypt(&self, x: usize) -> WObfBuffer<A> {
 		unsafe {
 			let mut buffer = WObfBuffer::<A>::uninit();
-			let data = self.data.as_slice();
+			let data = self.data.as_ref();
 			let src = data.as_ptr() as usize - data.len() * XREF_SHIFT;
 			let f: unsafe fn(&mut [u16], usize) = mem::transmute(ptr::read_volatile(&(wdecryptbuf as usize + x)) - x);
-			f(buffer.0.as_mut_slice(), src);
+			f(buffer.0.as_mut(), src);
 			buffer
 		}
 	}
 }
-impl<A: FixedSizeArray<u16>> fmt::Debug for WObfString<A> {
+impl<A: AsRef<[u16]> + AsMut<[u16]>> fmt::Debug for WObfString<A> {
 	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
 		self.decrypt(random!(usize) & 0xffff).fmt(f)
 	}
 }
-impl<A: FixedSizeArray<u16>> fmt::Display for WObfString<A> {
+impl<A: AsRef<[u16]> + AsMut<[u16]>> fmt::Display for WObfString<A> {
 	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
 		self.decrypt(random!(usize) & 0xffff).fmt(f)
 	}
@@ -287,29 +285,29 @@ unsafe fn wdecryptbuf(dest: &mut [u16], src: usize) {
 #[derive(Copy, Clone)]
 #[repr(transparent)]
 pub struct WObfBuffer<A>(A);
-impl<A: FixedSizeArray<u16>> WObfBuffer<A> {
+impl<A: AsRef<[u16]>> WObfBuffer<A> {
 	#[allow(deprecated)]
 	unsafe fn uninit() -> Self {
 		mem::uninitialized()
 	}
 	#[inline]
 	pub fn as_wide(&self) -> &[u16] {
-		self.0.as_slice()
+		self.0.as_ref()
 	}
 }
-impl<A: FixedSizeArray<u16>> ops::Deref for WObfBuffer<A> {
+impl<A: AsRef<[u16]>> ops::Deref for WObfBuffer<A> {
 	type Target = [u16];
 	#[inline]
 	fn deref(&self) -> &[u16] {
 		self.as_wide()
 	}
 }
-impl<A: FixedSizeArray<u16>> AsRef<[u16]> for WObfBuffer<A> {
+impl<A: AsRef<[u16]>> AsRef<[u16]> for WObfBuffer<A> {
 	fn as_ref(&self) -> &[u16] {
 		self.as_wide()
 	}
 }
-impl<A: FixedSizeArray<u16>> fmt::Debug for WObfBuffer<A> {
+impl<A: AsRef<[u16]>> fmt::Debug for WObfBuffer<A> {
 	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
 		use fmt::Write;
 		f.write_str("\"")?;
@@ -319,7 +317,7 @@ impl<A: FixedSizeArray<u16>> fmt::Debug for WObfBuffer<A> {
 		f.write_str("\"")
 	}
 }
-impl<A: FixedSizeArray<u16>> fmt::Display for WObfBuffer<A> {
+impl<A: AsRef<[u16]>> fmt::Display for WObfBuffer<A> {
 	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
 		use fmt::Write;
 		for chr in char::decode_utf16(self.as_wide().iter().cloned()) {

